src/agent/agent.c

```diff
@@ -2791,6 +2769,73 @@ static bool agent_connect(Agent *agent) {
                 }
         }
 
+        return 0;
+}
+
+static int agent_register_callback(sd_bus_message *m, void *userdata, UNUSED sd_bus_error *ret_error) {
+        Agent *agent = (Agent *) userdata;
+
+        if (agent->connection_state != AGENT_CONNECTION_STATE_CONNECTING) {
+                bc_log_error("Agent is not in CONNECTING state, dropping Register callback");
+                if (agent->register_call_slot != NULL) {
+                        sd_bus_slot_unrefp(&agent->register_call_slot);
+                }
+                return -EPERM;
+        }
+        int r = 0;
+        r = agent_process_register_callback(m, agent);
+        if (r < 0) {
+                agent->connection_state = AGENT_CONNECTION_STATE_RETRY;
+                return r;
+        }
+        return 0;
+}
+
+static bool agent_connect(Agent *agent) {
+        peer_bus_close(agent->peer_dbus);
+
+        bc_log_infof("Connecting to controller on %s", agent->orch_addr);
+        agent->connection_state = AGENT_CONNECTION_STATE_CONNECTING;
+
+        agent->peer_dbus = peer_bus_open(agent->event, "peer-bus-to-controller", agent->orch_addr);
+        if (agent->peer_dbus == NULL) {
+                bc_log_error("Failed to open peer dbus");
+                return false;
+        }
+
+        bus_socket_set_options(agent->peer_dbus, agent->peer_socket_options);
+
+        int r = sd_bus_add_object_vtable(
+                        agent->peer_dbus,
+                        NULL,
+                        INTERNAL_AGENT_OBJECT_PATH,
+                        INTERNAL_AGENT_INTERFACE,
+                        internal_agent_vtable,
+                        agent);
+        if (r < 0) {
+                bc_log_errorf("Failed to add agent vtable: %s", strerror(-r));
+                return false;
+        }
+
+        _cleanup_sd_bus_message_ sd_bus_message *bus_msg = NULL;
+        sd_bus_error error = SD_BUS_ERROR_NULL;
```

Comment:
Maybe change too `_cleanup_sd_bus_error_ sd_bus_error error = SD_BUS_ERROR_NULL;`

Reply:
If there is an error it is freed manually further, this was copied from the old code, but actually it is not needed anymore

Reply:
Good catch @ArtiomDivak 
Fixed